### PR TITLE
Add an updated hybrid retriever

### PR DESCRIPTION
**Pull request by palaniappan-r**
Changes and additions,
- Created a base chain class, derived all retrievers and chains from it
- Rewrote the existing vectorstore class
- Implemented a new hybrid search with multiple retrievers and a re-ranker
- Rewrote the POST endpoint to send sources and context separately
- Updated dependencies

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,24 +51,26 @@ Ensure you have `docker` and `docker-compose` installed in your system.
...
+
+    id5([Reranking]) -- top-n docs --> id6([LLM])
  
 ```
 
-By default,
-
 The backend will then be hosted at [http://0.0.0.0:8000](http://0.0.0.0:8000). 
 
 Open [http://0.0.0.0:8000/docs](http://0.0.0.0:8000/docs) for the API docs.
```
Comment (luarss) [inline README.md:76]: Insert some docs on how to use the Ruff to format and check code. 

Also are you planning it as a pre-commit / github CI? Or separate PR.

File: backend/src/vectorstores/faiss.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        embeddings_model_name: str,
+        distance_strategy: DistanceStrategy = DistanceStrategy.COSINE,
+        print_progress: bool = False,
+        debug: bool = False,
+        use_cuda: bool = False,
+    ):
+        self.embeddings_model_name = embeddings_model_name
+
+        model_kwargs = {"device": "cuda"} if use_cuda else {}
+        self.embedding_model = HuggingFaceEmbeddings(
```
Comment (luarss) [inline backend/src/vectorstores/faiss.py:25]: Write a method to handle different embeddings based on `embeddings_model_name`?

File: backend/src/chains/base_chain.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+from dotenv import load_dotenv
+
+from langchain.prompts import ChatPromptTemplate
+
+from langchain_core.output_parsers import StrOutputParser
+
+# from langchain_google_vertexai import ChatVertexAI
```
Comment (luarss) [inline backend/src/chains/base_chain.py:7]: Why comment?

File: .gitignore [hunk 4]
```diff
@@ -1,4 +1,6 @@
 __pycache__/
 .env
 data/
-backend/src/source_list.json
\ No newline at end of file
+backend/src/source_list.json
+.ruff_cache
+backend/src/secret.json
```
Comment (luarss) [inline .gitignore:6]: What is this used for?

File: README.md [hunk 5]
```diff
@@ -11,6 +11,10 @@ The current architecture uses certain retrieval techniques on OpenROAD documenta
 
 ## Setup
 
+### Building Manpages
+
+Build manpages as per the instructions [here](https://github.com/The-OpenROAD-Project/OpenROAD/tree/master/docs). Place the markdown files in `backend/data/markdown/manpages` before proceeding.
```
Comment (luarss) [inline README.md:16]: Maybe automate this step if possible? Or add a CLI option while building your docker image

File: README.md (hunk 1 above)
Comment (luarss) [reply README.md:76]: A `mypy` type checker would be good as well if time permits.

File: .gitignore (hunk 4 above)
Comment (palaniappan-r) [reply .gitignore:6]: It has the API keys required for using the newer models on Google's VertexAI platform. I'll add an option to use these models in the next PR.

**Review by luarss — DISMISSED** [review]

**Review by dralabeing — APPROVED** [review]

**Pull request merged**